utils/dwdparse/dwdparse_test.go

```diff
@@ -234,215 +233,204 @@ var dWDRules = []DWDirectiveRuleSpec{
-
-	// Examine each rule. If there is an error with the rule, return that as a failure.
-	// Otherwise, continue looking at all the rules to see if you can find a valid rule
-	// recording whether we found one. If the DWDirective matches a rule without other errors,
-	// return succes.
-	uniqueMap := make(map[string]bool)
-
-	for _, directive := range directiveList {
-		directiveMatchesARule := false // Anticipate failure
-		for i := range dwRules {
-			valid, err := ValidateDWDirective(dWDRules[i], directive, uniqueMap, failUnknownCommand)
-			if err != nil {
-				return err // Errors indicate parsing problems, reject directive
-			}
-
-			// The directive matched a rule
-			if valid {
-				directiveMatchesARule = true
-			}
-		}
+func TestDWParse(t *testing.T) {
+	for index, tt := range dwDirectiveTests {
+
+		err := Validate(dWDRules, tt.directiveList, func(int, DWDirectiveRuleSpec) {})
 
-		if !directiveMatchesARule {
-			return fmt.Errorf("invalid directive found: %s", directive)
+		if (tt.result == pass && err != nil) || (tt.result == fail && err == nil) {
+			t.Errorf("TestDWParse(%s)(%d): expect_valid(%v) err(%v)", tt.directiveList, index, tt.result, err)
 		}
 	}
+}
 
-	return nil
+// testCase defines an individual test case
+type testCase struct {
+	rules      []DWDirectiveRuleSpec
+	directives []string
+	result     bool
 }
 
-func TestDWParse(t *testing.T) {
-	for index, tt := range dwDirectiveTests {
-		validCommand := false
-		err := parsedw(t, tt.directiveList, dWDRules, tt.failUnknownCommand)
-		if err == nil {
-			validCommand = true
-		}
+// test provides a common method to test a series of test cases against a set of rules
+func test(t *testing.T, rules []DWDirectiveRuleSpec, tests []testCase) {
+	for index, tc := range tests {
+		err := Validate(rules, tc.directives, func(int, DWDirectiveRuleSpec) {})
 
-		if validCommand != tt.validCommand {
-			t.Errorf("TestDWParse(%s)(%d): allowUnsupportedCommand(%v) expect_valid(%v) err(%v)", tt.directiveList, index, tt.failUnknownCommand, tt.validCommand, err)
+		if (tc.result == pass && err != nil) || (tc.result == fail && err == nil) {
+			t.Errorf("test(%s)(%d): expected(%v) err(%v)", tc.directives, index, tc.result, err)
 		}
 	}
 }
+
+func TestUniqueWithin(t *testing.T) {
+	rules := []DWDirectiveRuleSpec{{
+		Command: "unique",
+		RuleDefs: []DWDirectiveRuleDef{{
+			Key:          "name",
+			Type:         "string",
+			IsRequired:   true,
+			UniqueWithin: "jobdw_name",
+		}},
+	}}
+
+	tests := []testCase{{
+		directives: []string{
+			"#DW unique name=uniqueName1 ",
+			"#DW unique name=uniqueName2 "},
+		result: pass,
+	}, {
+		directives: []string{
+			"#DW unique name=sameName",
+			"#DW unique name=sameName"},
+		result: fail,
+	}}
+
+	test(t, rules, tests)
+}
+
+func TestKeyRegexp(t *testing.T) {
+	rules := []DWDirectiveRuleSpec{{
+		Command: "regexp",
+		RuleDefs: []DWDirectiveRuleDef{{
+			Key:  "^(PREFIX_1_|PREFIX_2_)\\w+", // test a simple prefix regexp
```

Comment:
You should be able to use backticks here instead of double quotes and you won't have to escape the `\w`. Which begs the questions, will users need to escape it in the rule set?

Reply:
You're right.... damn let me change this.